x/stake/keeper/slash_test.go

```diff
@@ -0,0 +1,327 @@
+		ValidatorAddr: addrVals[1],
+		Shares:        sdk.NewRat(6),
+	}
+	keeper.SetDelegation(ctx, del)
+
+	// slash validator
+	ctx = ctx.WithBlockHeight(12)
+	oldPool := keeper.GetPool(ctx)
+	validator, found := keeper.GetValidatorByPubKey(ctx, pk)
+	require.True(t, found)
+	keeper.Slash(ctx, pk, 10, 10, fraction)
+
+	// read updating redelegation
+	rd, found = keeper.GetRedelegation(ctx, addrDels[0], addrVals[0], addrVals[1])
+	require.True(t, found)
+	// balance decreased
+	require.Equal(t, sdk.NewInt(3), rd.Balance.Amount)
+	// read updated pool
+	newPool := keeper.GetPool(ctx)
+	// bonded tokens burned
+	require.Equal(t, int64(4), oldPool.BondedTokens-newPool.BondedTokens)
+	// read updated validator
+	validator, found = keeper.GetValidatorByPubKey(ctx, pk)
+	// power decreased, but not by quite half, stake was bonded since
+	require.Equal(t, sdk.NewRat(8), validator.GetPower())
+}
+
+// tests Slash at a previous height with both an unbonding delegation and a redelegation
+func TestSlashBoth(t *testing.T) {
+	ctx, keeper, params, _, _ := setupHelper(t, 10)
+	fraction := sdk.NewRat(1).Quo(sdk.NewRat(2))
+
+	// set a redelegation
+	rdA := types.Redelegation{
+		DelegatorAddr:    addrDels[0],
+		ValidatorSrcAddr: addrVals[0],
+		ValidatorDstAddr: addrVals[1],
+		CreationHeight:   11,
+		MinTime:          0,
+		SharesSrc:        sdk.NewRat(6),
+		SharesDst:        sdk.NewRat(6),
+		InitialBalance:   sdk.NewCoin(params.BondDenom, 6),
+		Balance:          sdk.NewCoin(params.BondDenom, 6),
+	}
+	keeper.SetRedelegation(ctx, rdA)
+
+	// set the associated delegation
+	delA := types.Delegation{
+		DelegatorAddr: addrDels[0],
+		ValidatorAddr: addrVals[1],
+		Shares:        sdk.NewRat(6),
+	}
+	keeper.SetDelegation(ctx, delA)
+
+	// set an unbonding delegation
+	ubdA := types.UnbondingDelegation{
+		DelegatorAddr:  addrDels[0],
+		ValidatorAddr:  addrVals[0],
+		CreationHeight: 11,
+		MinTime:        0,
+		InitialBalance: sdk.NewCoin(params.BondDenom, 4),
+		Balance:        sdk.NewCoin(params.BondDenom, 4),
+	}
+	keeper.SetUnbondingDelegation(ctx, ubdA)
+
+	// slash validator
+	ctx = ctx.WithBlockHeight(12)
+	oldPool := keeper.GetPool(ctx)
+	validator, found := keeper.GetValidatorByPubKey(ctx, PKs[0])
+	require.True(t, found)
+	keeper.Slash(ctx, PKs[0], 10, 10, fraction)
+
+	// read updating redelegation
+	rdA, found = keeper.GetRedelegation(ctx, addrDels[0], addrVals[0], addrVals[1])
+	require.True(t, found)
+	// balance decreased
+	require.Equal(t, sdk.NewInt(3), rdA.Balance.Amount)
+	// read updated pool
+	newPool := keeper.GetPool(ctx)
+	// unbonding tokens burned
+	require.Equal(t, int64(2), oldPool.UnbondingTokens-newPool.UnbondingTokens)
+	// bonded tokens burned
+	require.Equal(t, int64(3), oldPool.BondedTokens-newPool.BondedTokens)
+	// read updated validator
+	validator, found = keeper.GetValidatorByPubKey(ctx, PKs[0])
+	// power not decreased, all stake was bonded since
+	require.Equal(t, sdk.NewRat(10), validator.GetPower())
+}
```

Comment:
I wouldn't mind seeing tests for recursive slashing as well as overslashing

Reply:
Will add overslashing tests. What do you mean by "recursive slashing"?

Reply:
I meant just multiple slashing of the same event on unbonding/redelegation objects